package/MDAnalysis/lib/src/clustering/ap.c

```diff
@@ -0,0 +1,269 @@
+/*
+ap.c --- C implementation of the Affinity Propagation clustering algorithm
+Copyright (C) 2014 Wouter Boomsma, Matteo Tiberti
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <float.h>
+
+/* Helper functions */
+
+inline int trmIndex(int row, int col) {  // array index for triangular matrix
+	return (row) > (col) ? ((row)+1)*(row)/2+(col) : ((col)+1)*(col)/2+(row);
+}
+
+inline int sqmIndex(int colsn, int row, int col) { // array index for square matrix
+	return row*colsn + col;
+}
+
+inline double pwmax(double x, double y) { //pairwise min
+	return x > y ? x : y;
+}
+
+inline double pwmin(double x, double y) { //pairwise max
+	return x < y ? x : y;
+}
+
+double min(double * values, int length) { //array min
+	double min = values[0];
+	for (int i=1;i<length;i++) {
+		if (values[i] < min) {
+			min = values[i];
+		}
+	}
+	return min;
+}
+
+double max(double * values, int length) { //array max
```

Comment:
Is this faster then the numpy functions? doesn't cython do something smart to find the maximum in a array?